React/Testing best practices.md

```diff
@@ -0,0 +1,610 @@
+
+## Setup
+
+First of all install jest
+
+```bash
+npm install --save-dev jest @types/jest
+```
+
+Add test script to `package.json`
+
+```js
+  "scripts": {
+    "test": "jest",
+    ...
+  }
+```
+
+Create `jest.config.js` in root of project
+
+```js
+module.exports = {
+  testPathIgnorePatterns: ["<rootDir>/.next/", "<rootDir>/node_modules/"],
+  moduleDirectories: ["node_modules", "src"],
+};
+```
+
+Add `preset` to `.babelrc`
+
+```js
+{
+  "presets": [
+    [
+      "next/babel"
+    ]
+  ]
+}
+```
+
+Next install react-test-library
+
+```bash
+npm i --save-dev @testing-library/react
+```
+
+### User events
+
+For mocking events [testing-library/user-event](https://github.com/testing-library/user-event) is used.
+
+```bash
+npm install --save-dev @testing-library/user-event
+```
+
+### Testing hooks
+
+For testing hooks [react-hooks-testing-library](https://github.com/testing-library/react-hooks-testing-library) is used.
+
+```bash
+npm install --save-dev @testing-library/react-hooks
+```
+
+### Add svg support
+
+```bash
+npm install --save-dev jest-svg-transformer
+```
+
+Add `transform` section to `jest.config.js`
+
+```js
+module.exports = {
+  ...
+  transform: {
+    '^.+\\.tsx?$': 'babel-jest',
+    '^.+\\.svg$': 'jest-svg-transformer',
+  },
+}
+```
+
+## Utils
+
+### Mock providers and store
+
+While testing we need to mock various providers like [`<ChakraProvider>`](https://chakra-ui.com/docs/getting-started#setup-provider/). To mock all provider we can create `./__tests__/test-utils.tsx` file where we will export [custom render](https://testing-library.com/docs/react-testing-library/setup/#custom-render) method with all providers.
+
+```tsx
+const AllProviders = ({ children }) => (
+  <ChakraProvider theme={theme}>{children}</ChakraProvider>
+);
+```
+
+Beside providers we need to mock [datx](https://datx.dev/) store
+
+```tsx
+export const StoreMockContext = React.createContext<AppCollection | null>(null);
+
+const withMockStore = (PageComponent: NextPage) => {
+  const WithMockStore: FC = (props) => {
+    const store: AppCollection = new AppCollection();
+
+    return (
+      <StoreMockContext.Provider value={store}>
+        <PageComponent {...props} />
+      </StoreMockContext.Provider>
+    );
+  };
+
+  return WithMockStore;
+};
+```
+
+Now while testing we can wrap a component in `<StoreMockContext.Consumer>` to get access to the store. Store is needed when we need to mock datx model with relationships.
+
+After mocking all providers and store all needs to be exported.
+
+```tsx
+const customRender = (ui: React.ReactElement, options?: any) =>
+  render(ui, { wrapper: withMockStore(AllProviders), ...options });
+
+// re-export everything
+export * from "@testing-library/react";
+
+// override render method
+export { customRender as render };
+```
+
+Now when testing components we won't import `render` from `@testing-library/react` but instead from `__tests__/test-utils.tsx` like this:
```

Comment:
```suggestion
Now, when testing components we won't import `render` from `__tests__/test-utils.tsx` instead of  `@testing-library/react`, like this:
```